interface/ceed-preconditioning.c

```diff
@@ -2579,7 +2576,8 @@ int CeedOperatorCreateFDMElementInverse(CeedOperator op, CeedOperator *fdm_inv,
 
   // Default interface implementation
   CeedCall(CeedOperatorGetCeed(op, &ceed));
-  CeedCall(CeedOperatorGetFallbackParentCeed(op, &ceed_parent));
+  CeedCall(CeedOperatorGetFallbackParent(op, &op_parent));
```

Comment:
This wouldn't be needed if the QF assembly logic wasn't changed. The intention was to use the fastest assembly process available any time the BuildOrAssemble function is called

Reply:
Addressing this comment along with https://github.com/CEED/libCEED/pull/1359#discussion_r1344715992 and https://github.com/CEED/libCEED/pull/1359#discussion_r1344714266 here:

The logic in `CeedOperatorLinearAssembleQFunctionBuildOrUpdate` used to go up one parent level and check for `LinearAssembleQFunctionUpdate`. Now, `CeedOperatorGetFallbackParent` goes all the way to the top (bottom? haha) of the inheritance tree to allow for multiple levels of overrides. In doing so this means `CeedOperatorLinearAssembleQFunctionBuildOrUpdate` is now written in the same way as `CeedOperatorLinearAssembleQFunction`: Only fallbacks are searched for definitions, no parent promotion. In addition to enabling this multi-level inheritance, it also helps reduce some confusion between what the two methods are doing. I think it can be very confusing to a user if calling `CeedOperatorLinearAssembleQFunctionBuildOrUpdate` instead of `CeedOperatorLinearAssembleQFunction` directly leads to a different backend being called.

Reply:
The goal with BuildOrUpdate was to build the QData as fast as possible without the caller having the write the same parent chasing all over the place though. All the backends share the same assembly data because of how the fallback is set up.

Reply:
I will refactor this to revert the need to perform the fallback parent search externally. That should go a long way towards improving your concern I think. The use case I have in mind, that breaks with the previous implementation, is that if we call `CeedOperatorLinearAssembleQFunctionBuildOrUpdate` having fallen back to `/cpu/self/ref/serial`, then only one backend level up was searched for having an `LinearAssembleQFunctionUpdate` definition. This would be incorrect if a backend which was two levels up in the fallback tree had a different definition. The code now handles this with `CeedOperatorGetFallbackParent`, but if no corresponding change was made to `CeedOperatorLinearAssembleQFunctionBuildOrUpdate` then this wouldn't work either since you can end up in infinite loops.

I will push a change shortly to have another go at this while addressing your concerns, give me a minute. Hopefully that will be an improvement.

Reply:
Originally it was intended that there would only be two levels, the fallback and its parent. The fallback would only be created and be responsible for any preconditioning implementation the parent is missing.

Reply:
Ah, this I did not realize. This isn't exactly the case anymore though, right? For CPU backends like `xsmm`, the first call to `CeedOperatorLinearAssemble` will go all the way back to `ref`, and then the parent points to `opt`, not `xsmm`, if I've traced this out correctly.

Reply:
There are two different parents. The inheritance of ceed functionality it separate from the fallback

Reply:
Right, I guess this is all not super important since the original reason why I was wanting to change this was to get access to the top level `Ceed` resource for creating a new `CeedTensorContract`. But that was a bad idea anyway and reverted. So I'll revert these changes here.

Reply:
This confusion is part of why I don't love the fallback mechanism.

---

`CeedSetDelegate` is our way to manage inheritance.

`/cpu/self/*/xsmm` inherits from `/cpu/self/*/opt` for all objects this backend doesn't implement (Vector, ElemRestriction, Basis, QFunction, and Operator)

`/cpu/self/*/opt` inherits from `/cpu/self/*/ref` for all objects this backend doesn't implement (Vector, ElemRestriction, Basis, and QFunction)

then `/cpu/self/*/ref` implements everything.

This does have the result that the `/xsmm` backend will create an `/opt` CeedOperator and use Vectors, Bases, and QFunction from `/ref` and a `TensorContract` from `/xsmm`

This is all-or-nothing for each object though. Either the backend fully implements the object or it  doesn't and we track back up the inheritance chain.

---

The idea with the fallback for CeedOperators only was to add partial inheritance just for CeedOperator functionality to preconditioning.

This way the `CeedOperatorApply` would take the faster path of the implementation at the bottom of the inheritance chain, but then a clone of the CeedOperator would be created by the fallback Ceed for the missing preconditioning related implementations. If there is no fallback Ceed given, then the interface level implementation is used instead for missing preconditioning related CeedOperator implimentations.

Probably, the code is not best organized to communicate that intent.

Reply:
Thanks a lot for the detailed description, it definitely does make sense. I've reverted all of the changes I had made with my incorrect interpretations, and even if the code isn't perfectly organized, it works so I have no complaints. Thanks a lot for your help!

Reply:
That description probably should live... somewhere? I'll try to find a place in the dev notes for it

Reply:
There's a brief few lines at the end of libCEEDdev.md which describes in less detail and without the usage context what you've said. I would say that copy and pasting you're few sentences into that `## Backend Inheritance` would be a great change.